Threaded emscripten support

JoeOsborn opened the pull request:
This patch (on top of emscripten_testing, not develop) allows for pthread compilation of mupen64.  Unfortunately as long as libco is still in use it must be built with ASYNC=1 as well as pthread=(POOL_SIZE).

ToadKing committed:
start on emscripten support
builds but runs with errors

m4xw committed:
save wip

warmenhoven committed:
parallel rsp/rdp on ios and tvos

m4xw committed:
First batch of fixes

m4xw committed:
Yeet pre-generated headers

m4xw committed:
Attempt android fix

m4xw committed:
save

m4xw committed:
More shenans

m4xw committed:
Reduce memory allocation size for emscripten

m4xw committed:
[Emscripten] Set save_state_in_background

warmenhoven committed:
Enable parallel-rdp on iOS/tvOS

m4xw committed:
git subrepo clone --force git@github.com:libretro-fork/libretro-common-mupen64plus-nx.git libretro-common/

subrepo:
  subdir:   "libretro-common"
  merged:   "f9f6ac61c"
upstream:
  origin:   "git@github.com:libretro-fork/libretro-common-mupen64plus-nx.git"
  branch:   "master"
  commit:   "f9f6ac61c"
git-subrepo:
  version:  "0.4.6"
  origin:   "???"
  commit:   "???"

m4xw committed:
Misc fixes for reinit

warmenhoven committed:
iOS/tvOS: change a few build flags to try to get a bit more performance

jonian committed:
Add support for RPi5 builds

LibretroAdmin committed:
Merge pull request #538 from warmenhoven/warmenhoven/pr/ios-flags

iOS/tvOS: change a few build flags to try to get a bit more performance

m4xw committed:
Some stability improvements

m4xw committed:
git subrepo clone --force --branch=rebase ../mupen64plus-core mupen64plus-core/

subrepo:
  subdir:   "mupen64plus-core"
  merged:   "efb46b8a1"
upstream:
  origin:   "../mupen64plus-core"
  branch:   "rebase"
  commit:   "efb46b8a1"
git-subrepo:
  version:  "0.4.6"
  origin:   "???"
  commit:   "???"

m4xw committed:
Update Ini Headers

m4xw committed:
Update precompiled headers

m4xw committed:
Fix branch after rebase

m4xw committed:
git subrepo push mupen64plus-core

subrepo:
  subdir:   "mupen64plus-core"
  merged:   "a362165c0"
upstream:
  origin:   "https://github.com/libretro/mupen64plus-core.git"
  branch:   "master"
  commit:   "a362165c0"
git-subrepo:
  version:  "0.4.6"
  origin:   "???"
  commit:   "???"

m4xw committed:
Bump ver

warmenhoven committed:
iOS/tvOS: move build flags inside makefile

warmenhoven committed:
osx: also move these flags into the makefile instead of gitlab-ci

warmenhoven committed:
fix leftover line

warmenhoven committed:
Additional neon optimization flag

m4xw committed:
Stub ConfigGetParamBool use

warmenhoven committed:
Fix gliden64 reinitialization on second playthrough

bslenul committed:
Parse "&" properly from GLideN64.custom.ini

m4xw committed:
git subrepo clone --force ../GLideN64 GLideN64/

subrepo:
  subdir:   "GLideN64"
  merged:   "ff0793c9c"
upstream:
  origin:   "../GLideN64"
  branch:   "develop"
  commit:   "ff0793c9c"
git-subrepo:
  version:  "0.4.6"
  origin:   "???"
  commit:   "???"

m4xw committed:
Fixes after GLideN64 rebase
Note: Removal of NoiseTex

m4xw committed:
[GLideN64]: Update INI

m4xw committed:
git subrepo push GLideN64

subrepo:
  subdir:   "GLideN64"
  merged:   "50604252b"
upstream:
  origin:   "git@github.com:libretro/GLideN64.git"
  branch:   "develop"
  commit:   "50604252b"
git-subrepo:
  version:  "0.4.6"
  origin:   "???"
  commit:   "???"

m4xw committed:
git subrepo clone --force --branch=develop ../GLideN64 GLideN64/

subrepo:
  subdir:   "GLideN64"
  merged:   "b33fa191b"
upstream:
  origin:   "../GLideN64"
  branch:   "develop"
  commit:   "b33fa191b"
git-subrepo:
  version:  "0.4.6"
  origin:   "https://github.com/ingydotnet/git-subrepo.git"
  commit:   "110b9eb"

m4xw committed:
Update GLideN64 INI

bslenul committed:
Add "EnableCopyColorFromRDRAM" core option

bslenul committed:
Apply controllers options on-the-fly

bslenul committed:
Hide "EnableCopyColorFromRDRAM" for non-GLideN64 plugins

m4xw committed:
git subrepo push mupen64plus-core

subrepo:
  subdir:   "mupen64plus-core"
  merged:   "06fde7f12"
upstream:
  origin:   "https://github.com/libretro/mupen64plus-core.git"
  branch:   "master"
  commit:   "06fde7f12"
git-subrepo:
  version:  "0.4.6"
  origin:   "???"
  commit:   "???"

m4xw committed:
git subrepo clone --force ../mupen64plus-core mupen64plus-core/

subrepo:
  subdir:   "mupen64plus-core"
  merged:   "ef94b0dc7"
upstream:
  origin:   "../mupen64plus-core"
  branch:   "master"
  commit:   "ef94b0dc7"
git-subrepo:
  version:  "0.4.6"
  origin:   "???"
  commit:   "???"

m4xw committed:
git subrepo pull mupen64plus-rsp-paraLLEl

subrepo:
  subdir:   "mupen64plus-rsp-paraLLEl"
  merged:   "fd28f47a9"
upstream:
  origin:   "git@github.com:libretro/parallel-rsp.git"
  branch:   "master"
  commit:   "fd28f47a9"
git-subrepo:
  version:  "0.4.6"
  origin:   "https://github.com/ingydotnet/git-subrepo.git"
  commit:   "110b9eb"

m4xw committed:
Yeet precompiled headers

m4xw committed:
Rely on HAVE_UNISTD_H

m4xw committed:
Add precompiled headers

m4xw committed:
Bump ver

JoeOsborn committed:
Threaded emscripten fixes

m4xw commented:
I need to resync my emscripten branch its a bit out of date again..

Coolcord committed:
Added 5120x1440 and 7680x2160 as 32:9 widescreen resolutions to libretro

m4xw committed:
Temp Backport Disable dual-source blending

JoshuaEagles committed:
Add Memory Map Setup (#545)

* add memory map setup

* move retroarch memory map setup to device.c

* update retroarch memory map setup to map address space properly

* add memory map flags for the retroarch memory maps

* refactor to map the entirety of KSEG0 and KSEG1

* finish mapping the whole KSEG1 address space and improve code readability

* remove the -1 for the romsize mapping

* set dd_rom_size when setting the pointer for it

* fix pointers to dd regs and rom

* adjust retroarch memory mapping to use the existing struct from init_device

* revise use of the other mappings struct and fix broken pointers

m4xw committed:
Add mediump precision specification for fragment shaders

m4xw committed:
Disable dual source blending in OpenGL context

orbea commented on Makefile:
```diff
@@ -526,7 +526,10 @@ else ifeq ($(platform), emscripten)
-
+   ifneq ($(pthread),0)
+     CPUFLAGS += -pthread
+     LDFLAGS += -lpthread
```
Drive by review:

The -pthread flag is both a linker flag and a cflag, you shouldn't use -lpthread directly and instead rely on the compiler to set it with -pthread.

warmenhoven committed:
apple: use audiotoolbox for resampling

warmenhoven committed:
apple: resetting the audio converter unnecessarily can lead to audio pops

pstef committed:
GLideN64 gfx context on GL context reset

GLideN64's BufferedDrawer uses persistent buffer mappings
(GL_MAP_PERSISTENT_BIT) which become invalid when the GL context is
destroyed during fullscreen toggle. RetroArch doesn't call
context_destroy() before recreating the context, so detect the
context recreation by checking if emulation is already running when
context_reset() is called.

pstef committed:
Detect freedreno by its "FDxxx" renderer string

and classify FD5xx as Adreno and FD6xx as Adreno_no_bugs for proper
workaround selection.

pstef committed:
Disable ext_fetch (GL_EXT_shader_framebuffer_fetch) on freedreno desktop OpenGL

as it causes rendering issues with N64 depth compare.

JoeOsborn committed:
Add binbashbanana xxhash fix

JoeOsborn committed:
Document afterplay-specific hack and avoid when emscripten pthreads are used

This fixes emscripten savestates under the condition ASYNC=1
PTHREAD=nonzero and the threaded task executor is in use.

JoeOsborn commented:
This is rebased on upstream/develop and adds a fix for savestates under emscripten-with-pthreads.
